EIPS/eip-6327.md

```diff
@@ -0,0 +1,316 @@
+        uint expiration1,
+        uint allhash1,
+        uint[8] memory proof2,
+        uint pwdhash2,
+        uint expiration2,
+        uint allhash2
+    ) public override {
+        uint nonce = nonceOf[msg.sender];
+
+        if (nonce == 0) {
+            //init password
+
+            pwdhashOf[msg.sender] = pwdhash2;
+            nonceOf[msg.sender] = 1;
+            verify(msg.sender, proof2, 0, expiration2, allhash2);
+        } else {
+            //reset password
+
+            // check old pwdhash
+            verify(msg.sender, proof1, 0, expiration1, allhash1);
+
+            // check new pwdhash
+            pwdhashOf[msg.sender] = pwdhash2;
+            verify(msg.sender, proof2, 0, expiration2, allhash2);
+        }
+
+        emit SetPassword(msg.sender, pwdhash2);
+    }
+
+    function verify(
+        address user,
+        uint[8] memory proof,
+        uint datahash,
+        uint expiration,
+        uint allhash
+    ) public override {
+        require(
+            block.timestamp < expiration,
+            "ZKPass::verify: expired"
+        );
+
+        uint pwdhash = pwdhashOf[user];
+        require(
+            pwdhash != 0,
+            "ZKPass::verify: user not exist"
+        );
+
+        uint nonce = nonceOf[user];
+        uint fullhash = uint(keccak256(abi.encodePacked(expiration, block.chainid, nonce, datahash))) / 8; // 256b->254b
+        require(
+            verifyProof(proof, pwdhash, fullhash, allhash),
+            "ZKPass::verify: verify proof fail"
+        );
+
+        nonceOf[user] = nonce + 1;
+
+        emit Verified(user, nonce);
+    }
+
+    /////////// util ////////////
+
+    function verifyProof(
+        uint[8] memory proof,
+        uint pwdhash,
+        uint fullhash, //254b
+        uint allhash
+    ) internal view returns (bool) {
+        return
+            verifier.verifyProof(
+                [proof[0], proof[1]],
+                [[proof[2], proof[3]], [proof[4], proof[5]]],
+                [proof[6], proof[7]],
+                [pwdhash, fullhash, allhash]
+            );
+    }
+}
+```
+
+verifier.sol is auto generated by snarkjs, the source code circuit.circom is below
+
+```javascript
+pragma circom 2.0.0;
+
+include "../../node_modules/circomlib/circuits/poseidon.circom";
+
+template Main() {
+    signal input in[3];
+    signal output out[3];
+
+    component poseidon1 = Poseidon(2);
+    component poseidon2 = Poseidon(2);
+
+    poseidon1.inputs[0] <== in[0];  //pwd
+    poseidon1.inputs[1] <== in[1];  //address
+    out[0] <== poseidon1.out; //pwdhash
+
+    poseidon2.inputs[0] <== poseidon1.out;
+    poseidon2.inputs[1] <== in[2]; //fullhash
+    out[1] <== in[2]; //fullhash
+    out[2] <== poseidon2.out; //allhash
+}
+
+component main = Main();
+```
+
+
+## Security Considerations
+
+Since the pwdhash is public, it is possible to be crack the password. We estimate the Poseidon hash rate of RTX3090 would be 100Mhash/s, this is the estimate of crack time:
```

Comment:
The EIP looks good as a draft. And thank you for this creative idea!

[Peer Review] The only technical concern I have is that in 2008-2009 the Bitcoin's hashrate globally is 10MHash/s, today it's 400EHash/s which is 4e13 times more. If the hash rate of Poseidon increases as fast in that rate, it will be cracked in 100seconds in 15th year.  116586246y = 3.6766638539e+15sec

Reply:
Thank you, thanks for the notice, we are following, if any better way, we will update